core/log/profiler_hook.cpp

```diff
@@ -0,0 +1,372 @@
+notice, this list of conditions and the following disclaimer in the
+documentation and/or other materials provided with the distribution.
+
+3. Neither the name of the copyright holder nor the names of its
+contributors may be used to endorse or promote products derived from
+this software without specific prior written permission.
+
+THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS
+IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED
+TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A
+PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+(INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+******************************<GINKGO LICENSE>*******************************/
+
+#include <ginkgo/core/log/profiler_hook.hpp>
+
+
+#include <memory>
+#include <mutex>
+#include <sstream>
+
+
+#include <ginkgo/core/base/name_demangling.hpp>
+#include <ginkgo/core/log/logger.hpp>
+#include <ginkgo/core/stop/criterion.hpp>
+
+
+#include "core/log/profiler_hook.hpp"
+
+
+namespace gko {
+namespace log {
+
+
+void ProfilerHook::on_allocation_started(const gko::Executor* exec,
+                                         const gko::size_type&) const
+{
+    this->maybe_synchronize(exec);
+    this->begin_hook_("allocate", profile_event_category::memory);
+}
+
+
+void ProfilerHook::on_allocation_completed(const gko::Executor* exec,
+                                           const gko::size_type&,
+                                           const gko::uintptr&) const
+{
+    this->maybe_synchronize(exec);
+    this->end_hook_("allocate", profile_event_category::memory);
+}
+
+
+void ProfilerHook::on_free_started(const gko::Executor* exec,
+                                   const gko::uintptr&) const
+{
+    this->maybe_synchronize(exec);
+    this->begin_hook_("free", profile_event_category::memory);
+}
+
+
+void ProfilerHook::on_free_completed(const gko::Executor* exec,
+                                     const gko::uintptr&) const
+{
+    this->maybe_synchronize(exec);
+    this->end_hook_("free", profile_event_category::memory);
+}
+
+
+void ProfilerHook::on_copy_started(const gko::Executor* from,
+                                   const gko::Executor* to, const gko::uintptr&,
+                                   const gko::uintptr&,
+                                   const gko::size_type&) const
+{
+    this->maybe_synchronize(from);
+    this->maybe_synchronize(to);
+    this->begin_hook_("copy", profile_event_category::operation);
```

Comment:
Would this also not qualify as a memory event ?

Reply:
some of our copies happen via kernels, so I think they are more similar to kernels than to allocation/free